code/WorkInProgress/TempEngine.dm

```diff
@@ -36,39 +123,161 @@
 		if(!input_starting_pressure)
 			return null
 
+		var/datum/gas_mixture/gas_input = air1
+		var/datum/gas_mixture/gas_output = air2
+
 		//Calculate necessary moles to transfer using PV = nRT
 		var/pressure_delta = (input_starting_pressure - output_starting_pressure)/2
 
 		// Check if fan/blower is required to overcome passive gate
-		if(input_starting_pressure < (output_starting_pressure+src.min_circ_pressure))
-			pressure_delta = src.min_circ_pressure
-			// P = dp q / μf, q ignored for simplification of system
-			var/total_pressure = (output_starting_pressure + pressure_delta - input_starting_pressure)
-			fan_power_draw = round((total_pressure) / src.fan_efficiency)
+		if(circulator_flags & BACKFLOW_PROTECTION)
+			if(input_starting_pressure < (output_starting_pressure+src.min_circ_pressure))
+				pressure_delta = src.min_circ_pressure
+				// P = dp q / μf, q ignored for simplification of system
+				var/total_pressure = (output_starting_pressure + pressure_delta - input_starting_pressure)
+				fan_power_draw = round((total_pressure) / src.fan_efficiency)
+		else if(pressure_delta < 0)
+			gas_input = air2
+			gas_output = air1
+
+		pressure_delta *= src.lube_boost
+
+		if(fan_power_draw)
+			if(src.status & NOPOWER)
+				src.last_pressure_delta = 0
+				return null
+			else src.use_power(fan_power_draw)
+
+		// Calculate and perform gas transfer from in to out
+		var/transfer_moles = abs(pressure_delta)*gas_output.volume/max(gas_input.temperature * R_IDEAL_GAS_EQUATION, 1) //Stop annoying runtime errors
+		src.last_pressure_delta = pressure_delta
+		var/datum/gas_mixture/removed = gas_input.remove(transfer_moles)
 
-			if(fan_power_draw)
-				if(src.status & NOPOWER)
-					src.last_pressure_delta = 0
-					return null
-				else src.use_power(fan_power_draw)
+		handle_reactions(removed)
 
+		// Leaks gas varient
+		if((circulator_flags & LEAKS_GAS ) && prob(5))
+			var/datum/gas_mixture/leaked = gas_input.remove_ratio(rand(2,8)*0.01)
+			src.audible_message("<span class='alert'>[src] makes a hissing sound.</span>")
+			if(leaked) loc.assume_air(leaked)
 
-		var/transfer_moles = abs(pressure_delta)*src.air2.volume/max(src.air1.temperature * R_IDEAL_GAS_EQUATION, 1) //Stop annoying runtime errors
-		src.last_pressure_delta = pressure_delta
+		if(src.network1) src.network1.update = 1
+		if(src.network2) src.network2.update = 1
```

Comment:
why not just `?.`

Reply:
Incorrectly assumed this would resolve to an assignment to null.  Slick syntax is slick.